utils/writeable/src/try_writeable.rs

```diff
@@ -0,0 +1,442 @@
+    ///
+    /// ```
+    /// # use writeable::TryWriteable;
+    /// # let my_writeable: Result<&str, &str> = Ok("");
+    /// # let mut sink = String::new();
+    /// let _ = my_writeable.try_write_to(&mut sink)?;
+    /// # Ok::<(), core::fmt::Error>(())
+    /// ```
+    ///
+    /// # Examples
+    ///
+    /// The following examples use `Result<&str, usize>`, which implements [`TryWriteable`].
+    ///
+    /// Success case:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let w: Result<&str, usize> = Ok("success");
+    /// let mut sink = String::new();
+    /// let result = w.try_write_to(&mut sink);
+    ///
+    /// assert_eq!(result, Ok(Ok(())));
+    /// assert_eq!(sink, "success");
+    /// ```
+    ///
+    /// Failure case:
+    ///
+    /// ```
+    /// use writeable::TryWriteable;
+    ///
+    /// let w: Result<&str, usize> = Err(44);
+    /// let mut sink = String::new();
+    /// let result = w.try_write_to(&mut sink);
+    ///
+    /// assert_eq!(result, Ok(Err(44)));
+    /// assert_eq!(sink, "44");
+    /// ```
+    fn try_write_to<W: fmt::Write + ?Sized>(
+        &self,
+        sink: &mut W,
+    ) -> Result<Result<(), Self::Error>, fmt::Error> {
+        self.try_write_to_parts(&mut helpers::CoreWriteAsPartsWrite(sink))
+    }
+
+    /// Writes the content of this writeable to a sink with parts (annotations).
+    ///
+    /// For more information, see:
+    ///
+    /// - [`TryWriteable::try_write_to()`] for the general behavior.
+    /// - [`TryWriteable`] for an example with parts.
+    /// - [`Part`] for more about parts.
+    fn try_write_to_parts<S: PartsWrite + ?Sized>(
+        &self,
+        sink: &mut S,
+    ) -> Result<Result<(), Self::Error>, fmt::Error>;
+
+    /// Returns a hint for the number of UTF-8 bytes that will be written to the sink.
+    ///
+    /// This function returns `Some(error)` _if and only if_ [`TryWriteable::try_write_to()`]
+    /// returns `Err`.
+    fn try_writeable_length_hint(&self) -> (LengthHint, Option<Self::Error>) {
+        // TODO: Discuss this function, its signature, and its behavior
+        (LengthHint::undefined(), None)
+    }
+
+    /// Writes the content of this writeable to a string.
+    ///
+    /// This function does not return a string in the failure case. If you need a replacement
+    /// string ("lossy mode"), use [`TryWriteable::try_write_to()`] instead.
+    fn try_write_to_string(&self) -> Result<Cow<str>, Self::Error> {
+        let (hint, hint_error) = self.try_writeable_length_hint();
+        if hint.is_zero() {
+            return Ok(Cow::Borrowed(""));
+        }
+        let mut output = String::with_capacity(hint.capacity());
+        let result = match self.try_write_to(&mut output) {
+            Ok(result) => result,
+            Err(core::fmt::Error) => {
+                debug_assert!(false, "String infallible");
+                Ok(())
+            }
+        };
```

Comment:
that debug assert is unreachable, you can appease the type checker more concisely

```suggestion
        let result = self.try_write_to(&mut output).unwrap_or(Ok(());
```

Reply:
I prefer being explicit about this being a `core::fmt::Error`, especially since there is a doubly nested result going on. I want to be clear that I'm not discarding the `Self::Error`.

Reply:
How about

```suggestion
        let result = self.try_write_to(&mut output).unwrap_or_else(|core::fmt::Error| Ok(()));
```